src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt

```diff
@@ -437,6 +438,32 @@ data class InboundLiquidityOutgoingPayment(
     override val amount: MilliSatoshi = fees
     override val completedAt: Long? = lockedAt
     val fundingFee: LiquidityAds.FundingFee = LiquidityAds.FundingFee(purchase.fees.total.toMilliSatoshi(), txId)
+    /**
+     * Even in the "from future htlc" case the mining fee corresponding to the previous channel output
+     * will be paid immediately from the channel balance, except if there is no channel.
+     *
+     * In the "from future htlc case", this inbound liquidity purchase is going to be followed by one or several [IncomingPayment.ReceivedWith.LightningPayment]
+     * with a non-null [LiquidityAds.FundingFee] where [LiquidityAds.FundingFee.fundingTxId] matches this [InboundLiquidityOutgoingPayment.txId].
+     * The sum of [LiquidityAds.FundingFee.amount] will match [InboundLiquidityOutgoingPayment.feePaidFromFutureHtlc].
+     */
+    val feePaidFromChannelBalance = when (purchase.paymentDetails.paymentType) {
+        is LiquidityAds.PaymentType.FromChannelBalance -> ChannelManagementFees(miningFee = miningFees, serviceFee = purchase.fees.serviceFee)
+        is LiquidityAds.PaymentType.FromChannelBalanceForFutureHtlc -> ChannelManagementFees(miningFee = miningFees, serviceFee = purchase.fees.serviceFee)
+        is LiquidityAds.PaymentType.FromFutureHtlc -> ChannelManagementFees(miningFee = miningFees - purchase.fees.miningFee, serviceFee = 0.sat)
+        is LiquidityAds.PaymentType.FromFutureHtlcWithPreimage -> ChannelManagementFees(miningFee = miningFees - purchase.fees.miningFee, serviceFee = 0.sat)
+        is LiquidityAds.PaymentType.Unknown -> TODO()
+    }
+    val feePaidFromFutureHtlc = when (purchase.paymentDetails.paymentType) {
+        is LiquidityAds.PaymentType.FromChannelBalance -> ChannelManagementFees(miningFee = 0.sat, serviceFee = 0.sat)
+        is LiquidityAds.PaymentType.FromChannelBalanceForFutureHtlc -> ChannelManagementFees(miningFee = 0.sat, serviceFee = 0.sat)
+        is LiquidityAds.PaymentType.FromFutureHtlc -> ChannelManagementFees(miningFee = purchase.fees.miningFee, serviceFee = purchase.fees.serviceFee)
+        is LiquidityAds.PaymentType.FromFutureHtlcWithPreimage -> ChannelManagementFees(miningFee = purchase.fees.miningFee, serviceFee = purchase.fees.serviceFee)
+        is LiquidityAds.PaymentType.Unknown -> TODO()
+    }
+    val feeCreditUsed = when (purchase) {
+        is LiquidityAds.Purchase.WithFeeCredit -> purchase.feeCreditUsed
+        else -> 0.msat
```

Comment:
Couldn't we add a specific TLV in the htlc to pass this info explicitly?

> it may be cumbersome because it needs to read multiple entries from the DB (the liquidity purchase and the associated

We already have to do that in Phoenix, to link automated liquidity events to incoming payments.

Reply:
Let's merge this PR and address this point in a follow up if necessary.

Reply:
Right, but could we find a way to have this information "statelessly" in the HTLCs, even if it introduces a bit of redundancy ? Now that I think of it, I will definitely need this level of precision for the csv audit, and being unable to process rows one by one is going to be painful.

Earlier I mentioned that we already had to read multiple entries in phoenix, but that's for UI stuff and we don't need absolute precision there.